src/sinks/datadog/traces/apm_stats/flusher.rs

```diff
@@ -0,0 +1,173 @@
+pub async fn flush_apm_stats_thread(
+    mut tripwire: Receiver<Sender<()>>,
+    client: HttpClient,
+    compression: Compression,
+    endpoint_configuration: DatadogTracesEndpointConfiguration,
+    aggregator: Arc<Mutex<Aggregator>>,
+) {
+    let sender = ApmStatsSender {
+        client,
+        compression,
+        endpoint_configuration,
+        aggregator,
+    };
+
+    // flush on the same interval as the stats buckets
+    let mut interval =
+        tokio::time::interval(std::time::Duration::from_nanos(BUCKET_DURATION_NANOSECONDS));
+
+    debug!("Starting APM stats flushing thread.");
+
+    loop {
+        tokio::select! {
+
+        _ = interval.tick() => {
+            // flush the oldest bucket from the cache to Datadog
+            sender.flush_apm_stats(false).await;
+        },
+        signal = &mut tripwire =>  match signal {
+            // sink has signaled us that the process is shutting down
+            Ok(sink_shutdown_ack_sender) => {
+
+                debug!("APM stats flushing thread received exit condition. Flushing remaining stats before exiting.");
+                sender.flush_apm_stats(true).await;
+
+                // signal the sink (who tripped the tripwire), that we are done flushing
+                let _ = sink_shutdown_ack_sender.send(());
+                break;
+            }
+            Err(_) => error!(message = "Tokio Sender unexpectedly dropped.", internal_log_rate_limit = true),
+        }
+        }
+    }
+}
+
+struct ApmStatsSender {
+    client: HttpClient,
+    compression: Compression,
+    endpoint_configuration: DatadogTracesEndpointConfiguration,
+    aggregator: Arc<Mutex<Aggregator>>,
+}
+
+impl ApmStatsSender {
+    async fn flush_apm_stats(&self, force: bool) {
+        // explicit scope to minimize duration that the Aggregator is locked.
+        if let Some((payload, api_key)) = {
+            let mut aggregator = self.aggregator.lock().unwrap();
+            let client_stats_payloads = aggregator.flush(force);
+
+            if client_stats_payloads.is_empty() {
+                // no sense proceeding if no payloads to flush
+                None
+            } else {
+                let payload = StatsPayload {
+                    agent_hostname: aggregator.get_agent_hostname(),
+                    agent_env: aggregator.get_agent_env(),
+                    stats: client_stats_payloads,
+                    agent_version: aggregator.get_agent_version(),
+                    client_computed: false,
+                };
+
+                Some((payload, aggregator.get_api_key()))
+            }
+        } {
+            if let Err(e) = self.compress_and_send(payload, api_key).await {
+                // TODO emit an internal `Error` event here, probably
+                error!(
+                    message = format!("Error while encoding and sending APM stats payloads: {}", e),
+                    internal_log_rate_limit = true,
+                );
```

Comment:
No, no! 😂 You're already able to do printf-style formatting with `tracing::error!` itself, and ideally the arguments should be reordered, per the [style guide](https://github.com/vectordotdev/vector/blob/master/STYLE.md#passing-in-the-event-message), like this:

```suggestion
                error!(
                    internal_log_rate_limit = true,
                    "Error while encoding and sending APM stats payloads: {}",
                    e,
                );
```

Reply:
Doh!
Yeah I still need to make this second nature.
Despite the work in the internal metrics audit... guess it's not the same as writing them from scratch.

Reply:
This in particular is news to me...

```
// Don't do this:
debug!(%client_id, message = "Client entered authentication phase.");
debug!(message = "Client entered authentication phase.", %client_id);
// Do this instead:
debug!(%client_id, "Client entered authentication phase.");
```

Lots of areas in the code not following the preferred approach :cold_sweat:

Reply:
Yeah, I was discussing this recently with Bruce who raised the same remark on the style guide PR. Most call sites have not been updated yet. I am wondering if we could do a sweep with comby to get them all at once 🤔

Reply:
Looking forward to learning more about comby :)

Reply:
Heh. I tried changing the order of the fields for an internal event I'm adding ... and the `check-events` script is actually enforcing the wrong approach :sweat_smile:  guess we will need to change that...

Reply:
I think Toby updated the check-events script to accept both formats as part of the style guide introduction: https://github.com/vectordotdev/vector/pull/12482#discussion_r978970914

Maybe there is a bug with it though 🤔

Reply:
Ah, looks like the changes need to be applied to the `scan_logs()` function as well.

Reply:
Yeah, feel free to ignore this particular PR nit for now if the script is giving you hell about it.

Reply:
:+1: I made the changes to the standalone `error!` calls, but will have to hold off on applying that to the new internal event added.